PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+		using BTree  = Impl::BTree<K, V, B>;
+		using KVERef = typename BTree::KVERef;
+
+		KVERef front;
+		KVERef back;
+
+		struct KV {
+			const K& key;
+			V& value;
+		};
+
+		KV next_unchecked() {
+			KVERef kv_ref = this->front;
+
+			if (kv_ref.idx < kv_ref.node_ref.node->len) {
+				front = KVERef{kv_ref.node_ref, kv_ref.idx + 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx >= kv_ref.node_ref.node->len);
+
+			front = BTree::first_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx + 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV prev_unchecked() {
+			KVERef kv_ref = this->back;
+
+			if (kv_ref.idx > 0) {
+				this->back = KVERef{kv_ref.node_ref, kv_ref.idx - 1};
+				return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+			}
+
+			do {
+				kv_ref = kv_ref.node_ref.ascend().parent;
+			} while (kv_ref.idx == 0);
+
+			this->back = BTree::last_leaf_edge(KVERef{kv_ref.node_ref, kv_ref.idx - 1}.descend());
+			return {kv_ref.node_ref.node->keys[kv_ref.idx], kv_ref.node_ref.node->values[kv_ref.idx]};
+		}
+
+		KV next() {
+			assert(!(front == back));
+			return next_unchecked();
+		}
+	};
+
+	//todo(vuko): docs
+	template<typename K, typename V, size_t Bfactor = 6>
+	class BTreeMap {
+		static_assert(Bfactor > 1, "B factor must be greater than 1. Consider using a classic binary tree if you need a lesser value.");
+		using BTree    = Impl::BTree<K, V, Bfactor>;
+		using Root     = typename BTree::Root;
+		using LeafNode = typename BTree::LeafNode;
+		using KVERef   = typename BTree::KVERef;
+		using NodeRef  = typename BTree::NodeRef;
+		class Entry;
+		class Iter;
+		class Keys;
+		class Values;
+	public:
+		static constexpr size_t B = Bfactor;
+
+	public:
+		BTreeMap() : root{new LeafNode(), 0}, len(0) {} //todo(vuko): should we delay the allocation until the first insertion?
+		BTreeMap(BTreeMap&& other) : root(other.root), len(other.len) { ::new(&other) BTreeMap(); } //note(vuko): due to the previous issue this allocates!
+		//todo(vuko): copy constructor? tbh I'd rather avoid implementing it
+		~BTreeMap() { this->Clear(); delete this->root.node; };
+
+	public:
+		BTreeMap& operator=(BTreeMap&& other) {
+			this->~BTreeMap();
+			this->root = other.root;
+			this->len = other.len;
+			::new(&other) BTreeMap();
+			return *this;
+		}
+
+	public:
+		Optional<V> Insert(const K&  key, const V&  value) { return this->insert(          key ,           value ); }
+		Optional<V> Insert(const K&  key,       V&& value) { return this->insert(          key , std::move(value)); }
+		Optional<V> Insert(      K&& key, const V&  value) { return this->insert(std::move(key),           value ); }
+		Optional<V> Insert(      K&& key,       V&& value) { return this->insert(std::move(key), std::move(value)); }
```

Comment:
This can be done using `BTreeMap::Entry`
```
if (auto entry = map.Entry(key)) {
	//examples:
	some_ref = entry.OccupiedGet();
	entry.OccupiedInsert(value_a); //in case we want to replace
} else {
	entry.VacantInsert(value_b); //in case want only insertions
}
```
Now that I think about it, `OccupiedReplace()` seems like a better name... Thoughts?